File: 06/byungkyu/1392.cpp
class Solution {
public:
    int pi[100005]={0,};
    void kmpinit(int n,string S){
        int i,j,k,a,b,c;
        a=1;b=0;
        while(a+b<n){
            if(S[a+b]==S[b]){
                pi[a+b]=b+1;
                b++;
            }
            else{
                if(b==0)a++;
                else{
                    a+=b-pi[b-1];
                    b=pi[b-1];
                }
            }
        }
    }

    string longestPrefix(string S) {
        kmpinit(S.length(),S);
        return S.substr(0,pi[S.length()-1]);
    }
};

File: 06/taehyuk/1392.py
class Solution:
    def longestPrefix(self, s: str) -> str:
        def getPartialMatch(N: str) -> list:
            m = len(N)
            pi = [0] * m
            j = 0 

            for i in range(1, m):
                print(i, j, N[i], N[j])
                while (j > 0 and N[i] != N[j]):
                    print(j)
                    j = pi[j - 1]
                
                if N[i] == N[j]:
                    j += 1
                    pi[i] = j
                else:
                    pi[i] = 0
            
            return pi
        
        pi = getPartialMatch(s)
        result = pi[-1]
        
        return s[:result]
